include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -157,28 +159,101 @@ struct FFT<2,true, Scalar, device_capabilities>
 
 // ---------------------------- Below pending --------------------------------------------------
 
-/*
-
-// then define 4,8,16 in terms of calling the FFT<2> and doing the special radix steps before/after
-template<uint16_t K, bool Inverse, class device_capabilities>
-struct FFT
+// Forward FFT
+template<uint32_t K, typename Scalar, class device_capabilities>
+struct FFT<K, false, Scalar, device_capabilities>
 {
-    template<typename Accessor, typename ShaderMemoryAccessor>
-    static void __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(ShaderMemoryAccessor) sharedmemAccessor)
+    template<typename Accessor, typename SharedMemoryAccessor>
+    static enable_if_t< (mpl::is_pot_v<K> && K > 2), void > __call(NBL_REF_ARG(Accessor) accessor, NBL_REF_ARG(SharedMemoryAccessor) sharedmemAccessor)
     {
-        if (!Inverse)
+        static const uint32_t virtualThreadCount = K >> 1;
+        static const uint16_t passes = mpl::log2<K>::value - 1;
+        uint32_t stride = K >> 1;
+        //[unroll(passes)]
+        for (uint16_t pass = 0; pass < passes; pass++)
         {
-           ... special steps ...
+            //[unroll(K/2)]
+            for (uint32_t virtualThread = 0; virtualThread < virtualThreadCount; virtualThread++)
+            {
+                const uint32_t virtualThreadID = (virtualThread << _NBL_HLSL_WORKGROUP_SIZE_LOG_2_) + SubgroupContiguousIndex();
+
+                const uint32_t lsb = virtualThread & (stride - 1);
+                const uint32_t loIx = ((virtualThread ^ lsb) << 1) | lsb;
+                const uint32_t hiIx = loIx | stride;
+                
+                complex_t<Scalar> lo = accessor.get(loIx << _NBL_HLSL_WORKGROUP_SIZE_LOG_2_);
+                complex_t<Scalar> hi = accessor.get(hiIx << _NBL_HLSL_WORKGROUP_SIZE_LOG_2_);
+                
+                hlsl::fft::DIF<Scalar>::radix2(hlsl::fft::twiddle<false,Scalar>(virtualThreadID & ((stride << _NBL_HLSL_WORKGROUP_SIZE_LOG_2_) - 1), stride << _NBL_HLSL_WORKGROUP_SIZE_LOG_2_),lo,hi);
```

Comment:
the indexing here is pretty hard-core too

Reply:
why not count `stride` properly ?

why not count everything with the <<WG_SZ_LOG2 already done?

Reply:
new math is good